nemo/collections/speechlm2/models/duplex_ear_tts.py

```diff
@@ -0,0 +1,1991 @@
+            # get context_hidden_state it is always one step behind current_subword_id
+            # for the first step uses the last step from warmup
+            context_hidden_state = self.embed_tokens(prev_subword_id)
+        else:
+            context_hidden_state = None
+
+        # force silence as next token
+        if self.cfg.get('inference_force_speech_silence_on_eos', True):
+            silence_codes = self.codec_silence_tokens.view(1, 1, -1).expand(prev_audio_tokens.shape)
+            prev_audio_tokens = torch.where(
+                current_subword_id.unsqueeze(-1) == self.text_eos_id,
+                silence_codes,  # silence
+                prev_audio_tokens,  # keep original
+            )
+
+        # get subword_ids
+        inputs = {
+            "code": prev_audio_tokens,
+            "context_hidden_state": context_hidden_state,
+            "subword_ids": current_subword_id,
+            "subword_mask": current_subword_mask,
+            "past_key_values": past_key_values,
+            "use_cache": True,
+            "guidance_enabled": guidance_enabled,
+            "generation_config": generation_config,
+            "ignore_eos_flag_stop": ignore_eos_flag_stop,
+        }
+
+        outputs = self.tts_model(**inputs)
+
+        return outputs["codes"], outputs["past_key_values"]
+
+    @torch.no_grad()
+    def decode_one_audio_step(self, gen_audio_codes_history, number_prev_tokens=None):
+        """
+        Decodes one step of generated audio codec tokens to raw waveform.
+
+        Args:
+            gen_audio_codes_history (torch.Tensor): Audio tokens history, shape (B, T, C).
+            number_prev_tokens (int, optional): Number of previous tokens to decode, for incremental decoding.
+
+        Returns:
+            Tuple[torch.Tensor, torch.Tensor]:
+                - audio_pred_cur_step: Latest decoded waveform chunk, shape (B, wav_to_token_ratio).
+                - audio_len: Lengths (number of samples), shape (B,).
+        """
+        with fp32_precision(), torch.no_grad():
+            if number_prev_tokens:
+                gen_audio_codes_history = gen_audio_codes_history[:, -number_prev_tokens:]
+
+            gen_audio_codes_history = replace_control_speech_codes(
+                gen_audio_codes_history, self._control_codes, self.codec_silence_tokens
+            )
+            gen_audio_codes_lens = torch.tensor(
+                [gen_audio_codes_history.size(1)] * gen_audio_codes_history.size(0), device=self.device
+            )
+            audio_pred, audio_len = self.audio_codec.decode(gen_audio_codes_history, gen_audio_codes_lens)
+
+        # return only the current/lastest audio chunk
+        audio_pred_cur_step = audio_pred.squeeze(1)[:, -self.audio_codec.config.wav_to_token_ratio :]
+        audio_len[:] = self.audio_codec.config.wav_to_token_ratio
+        return audio_pred_cur_step, audio_len
+
+    @torch.no_grad()
+    def offline_inference(
```

Comment:
Why isn't this function used in the inference script rather than `trainer.validate`?

Reply:
This function is used during validation and can also be called from an external script. However, for now, we are only planning to add the validation recipe, since the full inference pipeline (including integration with S2T) will be implemented in NIVA.